core/drand_beacon.go

```diff
@@ -434,6 +447,98 @@ func (bp *BeaconProcess) newMetadata() *common.Metadata {
 	return metadata
 }
 
+var errNoRoundInPeers = errors.New("could not find round")
+
+func (bp *BeaconProcess) storeCurrentFromPeerNetwork(ctx context.Context, store chain.Store) error {
+	clkNow := bp.opts.clock.Now().Unix()
+	currentRound := chain.CurrentRound(clkNow, bp.group.Period, bp.group.GenesisTime)
+	if currentRound < 1 {
+		// We cannot sync the initial round.
+		// Assume this is a fresh start
+		return nil
+	}
+
+	peers := bp.computePeers(bp.group.Nodes)
+	previousRound, err := bp.loadBeaconFromPeers(ctx, currentRound, peers)
+	if err != nil {
+		return err
+	}
+
+	verif := chain.NewVerifier(bp.group.Scheme)
+	err = verif.VerifyBeacon(previousRound, bp.group.PublicKey.Key())
+	if err != nil {
+		bp.log.Errorw("failed to verify beacon", "err", err)
+		return err
+	}
+
+	err = store.Put(ctx, &previousRound)
+	if err != nil {
+		bp.log.Errorw("failed to store beacon", "err", err)
+	}
+	return err
+}
+
+func (bp *BeaconProcess) loadBeaconFromPeers(ctx context.Context, targetRound uint64, peers []net.Peer) (chain.Beacon, error) {
+	found := false
+	round := chain.Beacon{}
+
+	setMtx := sync.Mutex{}
+
+	findCtx, findCancel := context.WithCancel(ctx)
+	defer findCancel()
+
+	for _, peer := range peers {
+		peer := peer
+
+		go func() {
+			r, err := bp.privGateway.PublicRand(findCtx, peer, &drand.PublicRandRequest{
+				Round:    targetRound,
+				Metadata: bp.newMetadata(),
+			})
+			if err != nil {
+				bp.log.Errorw("failed to get rand value from peer", "round", targetRound, "err", err, "peer", peer.Address())
+				return
+			}
+
+			select {
+			case <-findCtx.Done():
+				return
+			default:
+			}
+
+			findCancel()
+
+			setMtx.Lock()
+			defer setMtx.Unlock()
+
+			found = true
+			round.PreviousSig = r.PreviousSignature
+			round.Round = r.Round
+			round.Signature = r.Signature
+		}()
+		break
+	}
+
+	if !found {
```

Comment:
Why use `found` and mutexes instead of a channel?

Reply:
Fixed